Architecture.md

```diff
@@ -71,6 +71,116 @@
+    - If the client is not authenticated, the connection is closed
+- The git server checks if the client is authorized to push to the repository by checking if the client is the owner of the repository
+    - If the client is not authorized, the connection is closed
+- The git server executes the command `git-receive-pack` which is used to push to a repository
+
+```mermaid
+sequenceDiagram
+    participant Client
+    participant GitServer
+    participant Database
+    participant HostMachine
+
+    Client->>GitServer: git-send-pack repository_name (git push)
+    GitServer->>Database: Is the ssh key in the database?
+    Database-->>GitServer: result
+    alt Authenticated
+        GitServer->>Database: Is the ssh key linked to the repository ?
+        Database-->>GitServer: result
+        alt Authorized
+            GitServer->>HostMachine: Execute git-receive-pack
+            HostMachine-->>GitServer: Push Successful (from host)
+            GitServer-->>Client: Push Successful (continue)
+        else Not Authorized
+            GitServer-->>Client: Connection Closed (Not Authorized)
+        end
+    else Not Authenticated
+        GitServer-->>Client: Connection Closed (Not Authenticated)
+    end
+```
+
+##### Building images
+
+Once the code is finally pushed to the repository, it needs to be built in order to be deployed. Since the strategy we chose is [Docker](https://www.docker.com/), the code needs to be built into a docker image.
+
+In that essence, since the client code can be written in any language, we need to be able to build the code in any language. To do so, we use [buildpacks](https://buildpacks.io/), which is a tool to build code in any language into a docker image. It is used by [heroku](https://www.heroku.com/) to build their applications.
+
+How can we trigger the build of the image ? For a first version, we chose to use [git hooks](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks) to trigger a bash script which will build the image on the host machine. This is not scalable, as the build is done on the host machine. It also poses questions about security risks since our ssh server needs to be able to access the docker daemon on the host machine. But it is a good first step to test the concept.
+
+Some builders might not be able to build the code without configuration, hence, we need to be able to configure the buildpacks. To do so, the client can store a file called `buildpacks.json` in the `paastech` directory which will allow the user to configure the buildpacks. The file is a json file, which contains the configuration of the buildpacks. It is then used by the git hook to configure buildpacks. Here's the structure of the file:
+
+```yaml
+buildpacks:
+  path: "."
+  builder: "paketobuildpacks/builder:full"
+  buildpack: "paketo-buildpacks/web-servers"
+  env:
+    - "BP_WEB_SERVER=nginx"
+    - "BP_WEB_SERVER_ROOT=dist"
+}
+```
```

Comment:
You indicate above that the file is a JSON, but you display a YAML file. Also, the last line is a closing bracket with no opening bracket whatsoever. Can you please fix this?

Reply:
fixed: d1fa38001144e77236b4d31dfa45aa1023afd45b